### PR TITLE
Position axis outside of canvas instead of setting viewport

### PR DESCRIPTION
**Pull request by axelboc**
I think it will make more sense if you look at the files from bottom to top... and if you inspect the DOM yourself.

### CODE REVIEW

File: src/h5web/visualizations/heatmap/HeatmapVis.module.css [hunk 1]
```diff
@@ -14,24 +14,32 @@
 
 .heatmap {
   flex: none;
+  box-sizing: content-box;
```
Comment (axelboc) [inline src/h5web/visualizations/heatmap/HeatmapVis.module.css:17]: This avoids having to add the axis offsets to the width and height.

File: src/h5web/visualizations/heatmap/hooks.ts [hunk 2]
```diff
@@ -42,8 +42,10 @@ export function useHeatmapSize<T>(
   return [
     wrapperRef,
     {
-      width: meshWidth + leftAxisWidth,
-      height: meshHeight + bottomAxisHeight,
+      width: meshWidth,
+      height: meshHeight,
+      paddingBottom: bottomAxisHeight,
+      paddingLeft: leftAxisWidth,
```
Comment (axelboc) [inline src/h5web/visualizations/heatmap/hooks.ts:48]: This is the most important bit: the width and height of the `heatmap` element in `HeatmapVis` now match the width and height of the canvas and of the mesh. It uses padding to leave space for the axis. Not that the CSS adds `box-sizing: content-box`, so the total width of the `heatmap` element ends up being `width + leftAxisWidth`.

File: src/h5web/visualizations/heatmap/Mesh.tsx [hunk 3]
```diff
@@ -1,33 +1,18 @@
...
-     * Set viewport of WebGL renderer to allow camera-related computations to be performed
-     * as if axis didn't exist and mesh filled entire canvas.
-     */
-    gl.setViewport(
-      leftAxisWidth,
-      bottomAxisHeight,
-      width - leftAxisWidth,
-      height - bottomAxisHeight
-    );
-  });
```
Comment (axelboc) [inline src/h5web/visualizations/heatmap/Mesh.tsx:30]: No more viewport. The mesh is rendered on the entire canvas.

File: src/h5web/visualizations/heatmap/HeatmapVis.tsx [hunk 4]
```diff
@@ -38,18 +38,16 @@ function HeatmapVis(props: Props): JSX.Element {
   return (
     <div className={styles.root}>
       <div ref={mapAreaRef} className={styles.mapArea}>
-        {heatmapSize && (
-          <div className={styles.heatmap} style={heatmapSize}>
-            <Canvas orthographic invalidateFrameloop>
+        {heatmapStyles && (
+          <div className={styles.heatmap} style={heatmapStyles}>
+            <Canvas
+              className={styles.canvasWrapper}
```
Comment (axelboc) [inline src/h5web/visualizations/heatmap/HeatmapVis.tsx:44]: Just adding a classname here to simplify selectors in CSS (`.heatmap > div` becomes `.canvasWrapper`)

File: src/h5web/visualizations/heatmap/HeatmapVis.tsx [hunk 5]
```diff
@@ -38,18 +38,16 @@ function HeatmapVis(props: Props): JSX.Element {
   return (
     <div className={styles.root}>
       <div ref={mapAreaRef} className={styles.mapArea}>
-        {heatmapSize && (
-          <div className={styles.heatmap} style={heatmapSize}>
-            <Canvas orthographic invalidateFrameloop>
+        {heatmapStyles && (
+          <div className={styles.heatmap} style={heatmapStyles}>
```
Comment (axelboc) [inline src/h5web/visualizations/heatmap/HeatmapVis.tsx:42]: Note that the styles here now include some padding for the axis.

File: src/h5web/visualizations/heatmap/HeatmapVis.module.css [hunk 6]
```diff
@@ -14,24 +14,32 @@
...
+}
+
+/* R3F's root element */
+.canvasWrapper {
+  overflow: visible !important; /* show axis */
```
Comment (axelboc) [inline src/h5web/visualizations/heatmap/HeatmapVis.module.css:22]: The `canvasWrapper` no longer includes the size of the axis, so we have to make overflow visible. This has a side-effect which requires setting `pointer-events: none` below.

File: src/h5web/visualizations/heatmap/AxisGrid.tsx [hunk 7]
```diff
@@ -54,16 +55,17 @@ function AxisGrid(props: Props): JSX.Element {
...
-          gridTemplateColumns: `${axisOffsets[0]}px 1fr`,
-          gridTemplateRows: `1fr ${axisOffsets[1]}px`,
-        }}
-      >
+    <Dom
+      className={styles.axisGrid}
+      style={{
+        width: width + leftAxisWidth,
+        height: height + bottomAxisHeight,
+        left: -leftAxisWidth,
```
Comment (axelboc) [inline src/h5web/visualizations/heatmap/AxisGrid.tsx:63]: The axis grid is now inside the canvas, which no longer includes the size of the axis, so we need to increase the grid's size and offset it to give the axis some space and position them properly.

File: src/h5web/visualizations/heatmap/HeatmapVis.tsx [hunk 8]
```diff
@@ -1,14 +1,14 @@
 import { Canvas } from 'react-three-fiber';
 import React, { useEffect, useMemo } from 'react';
-import Mesh from './Mesh';
 import { computeTextureData } from './utils';
 import styles from './HeatmapVis.module.css';
 import ColorBar from './ColorBar';
 import ColorMapSelector from './ColorMapSelector';
 import LogScaleToggler from './LogScaleToggler';
 import { useHeatmapState, useHeatmapActions } from './store';
+import { useHeatmapSize as useHeatmapStyles } from './hooks';
```
Comment (loichuder) [inline src/h5web/visualizations/heatmap/HeatmapVis.tsx:9]: You may want to rename `useHeatmapSize` in `useHeatmapStyles` in the `hooks` file as well

File: src/h5web/visualizations/heatmap/hooks.ts (hunk 2 above)
Comment (axelboc) [reply src/h5web/visualizations/heatmap/hooks.ts:48]: I moved the `content-box` style here even though it's static, so the code is more self-documenting.

File: src/h5web/visualizations/heatmap/HeatmapVis.module.css (hunk 6 above)
Comment (axelboc) [reply src/h5web/visualizations/heatmap/HeatmapVis.module.css]: I've reviewed the comment here.

File: src/h5web/visualizations/heatmap/AxisGrid.tsx (hunk 7 above)
Comment (axelboc) [reply src/h5web/visualizations/heatmap/AxisGrid.tsx:64]: I've added a comment here.

File: src/h5web/visualizations/heatmap/HeatmapVis.module.css [hunk 9]
```diff
@@ -16,22 +16,26 @@
   flex: none;
 }
 
-/* R3F's <Dom /> element*/
-.heatmap > div > div {
+/* R3F's root element */
+.canvasWrapper {
+  overflow: visible !important; /* show child axis grid, which is bigger than canvas */
```
Comment (axelboc) [inline src/h5web/visualizations/heatmap/HeatmapVis.module.css:21]: clearer comment here

File: src/h5web/visualizations/heatmap/hooks.ts [hunk 10]
```diff
@@ -29,21 +29,24 @@ export function useHeatmapSize<T>(
...
-  // Determine how to compute mesh size to fit available space while maintaining aspect ratio
+  // Determine how to compute canvas size to fit available space while maintaining aspect ratio
   const shouldAdjustWidth = availableWidth >= availableHeight * aspectRatio;
 
-  const meshWidth = shouldAdjustWidth
+  const canvasWidth = shouldAdjustWidth
     ? availableHeight * aspectRatio
     : availableWidth;
-  const meshHeight = shouldAdjustWidth
+  const canvasHeight = shouldAdjustWidth
```
Comment (axelboc) [inline src/h5web/visualizations/heatmap/hooks.ts:38]: I've renamed the variables here to make it clear that the styles are computed based on the size of the canvas.

File: src/h5web/visualizations/heatmap/hooks.ts [hunk 11]
```diff
@@ -19,31 +21,41 @@ export function useHeatmapSize<T>(
...
-    return [wrapperRef, { width, height }];
+    return [
+      wrapperRef,
+      {
+        width,
+        height,
+        paddingBottom: bottomAxisHeight,
+        paddingLeft: leftAxisWidth,
+      },
+    ];
```
Comment (axelboc) [inline src/h5web/visualizations/heatmap/hooks.ts:32]: J'avais oublié de mettre du padding ici...

**Pull request merged**